pyrdp/parser/rdp/ntlmssp.py

```diff
@@ -86,3 +93,118 @@ def parseNTLMSSPAuthenticate(self, data: bytes, stream: BytesIO) -> NTLMSSPAuthe
 
         return NTLMSSPAuthenticatePDU(user, domain, proof, response)
+
+    def parseNTLMSSPTSRequest(self, data: bytes, stream: BytesIO) -> NTLMSSPTSRequestPDU:
+        if not ber.readUniversalTag(stream, ber.Tag.BER_TAG_SEQUENCE, True):
+            raise UnknownPDUTypeError("Invalid BER tag (%d expected)" % ber.Tag.BER_TAG_SEQUENCE)
+
+        length = ber.readLength(stream)
+        if length > len(stream.getvalue()):
+            raise ParsingError("Invalid size for TSRequest (got %d, %d bytes left)" % (length, len(stream.getvalue())))
+
+        version = None
+        negoTokens = None
+
+        # [0] version
+        if not ber.readContextualTag(stream, 0, True):
+            return NTLMSSPTSRequestPDU(version, negoTokens, data)
+        version = ber.readInteger(stream)
+
+        # [1] negoTokens
+        if not ber.readContextualTag(stream, 1, True):
+            return NTLMSSPTSRequestPDU(version, negoTokens, data)
+        ber.readUniversalTag(stream, ber.Tag.BER_TAG_SEQUENCE, True)  # SEQUENCE OF NegoDataItem
+        ber.readLength(stream)
+        ber.readUniversalTag(stream, ber.Tag.BER_TAG_SEQUENCE, True)  # NegoDataItem
+        ber.readLength(stream)
+        ber.readContextualTag(stream, 0, True)
+
+        negoTokens = BytesIO(ber.readOctetString(stream))            # NegoData
+        return NTLMSSPTSRequestPDU(version, negoTokens)
+
+    def parseNTLMSSPChallengePayload(self, data: bytes, stream: BytesIO, workstationLen: int) -> NTLMSSPChallengePayloadPDU:
+        workstation = stream.read(workstationLen)
+        return NTLMSSPChallengePayloadPDU(workstation)
+
+    def writeNTLMSSPChallenge(self, workstation: str, serverChallenge: bytes) -> bytes:
+        stream = BytesIO()
+        substream = BytesIO()
+
+        workstation = workstation.encode('utf-16le')
+        nameLen = len(workstation)
+        pairsLen = self.writeNTLMSSPChallengePayload(substream, workstation)
+
+        substream.write(b'NTLMSSP\x00')
+        Uint32LE.pack(NTLMSSPMessageType.CHALLENGE_MESSAGE, substream)
+        Uint16LE.pack(nameLen, substream)
+        Uint16LE.pack(nameLen, substream)
+        Uint32LE.pack(NTLMSSPChallengeType.WORKSTATION_BUFFER_OFFSET, substream)
+        Uint32LE.pack(NTLMSSPChallengeType.NEGOTIATE_FLAGS, substream)
+        substream.write(serverChallenge)
+        Uint64LE.pack(0, substream)
+        Uint16LE.pack(pairsLen, substream)
+        Uint16LE.pack(pairsLen, substream)
+        Uint32LE.pack(NTLMSSPChallengeType.WORKSTATION_BUFFER_OFFSET + nameLen, substream)
+        Uint8.pack(NTLMSSPChallengeVersion.NEG_PROD_MAJOR_VERSION_HIGH, substream)
+        Uint8.pack(NTLMSSPChallengeVersion.NEG_PROD_MINOR_VERSION_LOW, substream)
+        Uint16LE.pack(NTLMSSPChallengeVersion.NEG_PROD_VERSION_BUILT, substream)
+        Uint8.pack(0, substream)
+        Uint8.pack(0, substream)
+        Uint8.pack(0, substream)
+        Uint8.pack(NTLMSSPChallengeVersion.NEG_NTLM_REVISION_CURRENT, substream)
```

Comment:
Where is that format documented? pairsLen, nameLen is written so often, I think there's something here that is unclear.

Could we do something similar to this?
```python
        domainLength = len(domain) - nullByteCount * unicodeMultiplier
        usernameLength = len(username) - nullByteCount * unicodeMultiplier
        passwordLength = len(password) - nullByteCount * unicodeMultiplier
        alternateShellLength = len(alternateShell) - nullByteCount * unicodeMultiplier
        workingDirLength = len(workingDir) - nullByteCount * unicodeMultiplier

        stream.write(Uint16LE.pack(domainLength))
        stream.write(Uint16LE.pack(usernameLength))
        stream.write(Uint16LE.pack(passwordLength))
        stream.write(Uint16LE.pack(alternateShellLength))
        stream.write(Uint16LE.pack(workingDirLength))
        stream.write(domain)
        stream.write(username)
        stream.write(password)
        stream.write(alternateShell)
        stream.write(workingDir)
```

Reply:
The documentation can be found in the [PDU](https://github.com/lubiedo/pyrdp/blob/issue-358/pyrdp/pdu/rdp/ntlmssp.py#L30-L37) but here's the link: https://docs.microsoft.com/en-us/openspecs/windows_protocols/ms-nlmp/801a4681-8809-4be9-ab0d-61dcfe762786

This is an specific struct that I'm trying to follow verbatim for better debugging and clarity. The reason why `nameLen` and `pairsLen` is written twice is because the first write refers to the length and the second to the max length, which are same in this case. This is an excerpt of the documentation, in this case for `nameLen`:

```
§ TargetNameLen (2 bytes): A 16-bit unsigned integer that defines the size, in bytes, of TargetName in Payload.
§ TargetNameMaxLen (2 bytes): A 16-bit unsigned integer that SHOULD be set to the value of TargetNameLen and MUST be ignored on receipt.
```

Reply:
This length, maxLength but is the same deal is so confusing. Not your fault, that's on Microsoft. Add that link to the method documentation and it's fine.